### PR TITLE
Fix Heroku Deployment

### PR DESCRIPTION
**Pull request by Reimirno**
Created a new staging environment in rails (basically the same to the production environment as close as possible, but with email sender error suppressed and database name changed etc.) This is meant for staging in Heroku so it staging environment does not do things that only production app should do (sending email to clients etc).

I ended up didn't changing anything in Procfile because: 

The right thing to do (for heroku) is just to:
- For the first deployment, comment out release command `db:prepare` and then deploy. Then, run `db:schema:load && bin/rails db:seed` manually on console.
- For all subsequent deployment, we uncomment release command `db:prepare` to allow it to run on subsequent deployments. Not doing so will leads to problems like when we have schema changes (when we add features, for example) that is not applied automatically... then you have to manually run `db:migrate` or `db:prepare` (if you have a db setup and schema loaded prepare basically just do migration). 

Can we simply replace the release command in Procfile to be `db:schema:load && bin/rails db:seed`? No. This will fail all subsequent runs because `db:schema:load` is considered destructive to database and running it a second time will be blocked by rails.

So that is why I ended up uncommenting the original `db:prepare`. Really, making sure `db:prepare` works in all cases (eliminating the edge case of first deployment) is the right thing to do, instead of swapping out command. I am not sure why `db:prepare` failed on the first deployment. Based on logs `db:prepare` on first deployment seems to fail at the migration stage, complaining either missing version number on migration file or missing table.

### CODE REVIEW

**Comment by Reimirno** [comment]
I will squash these into one commit if you think these are fine since essentially all the commits besides the one creating staging environment are just juggling the procfile and updating readme.

File: config/cable.yml [hunk 1]
```diff
@@ -8,3 +8,8 @@ production:
   adapter: redis
   url: <%= ENV.fetch("REDIS_URL") { "redis://localhost:6379/1" } %>
   channel_prefix: bjc_teacher_tracker_production
+
+staging:
+  adapter: redis
+  url: <%= ENV.fetch("REDIS_URL") { "redis://localhost:6379/1" } %>
+  channel_prefix: bjc_teacher_tracker_production
```
Comment (cycomachead) [inline config/cable.yml:15]: This name should probably be updated.

File: config/database.yml [hunk 2]
```diff
@@ -23,4 +23,10 @@ production:
   adapter: postgresql
   pool: 5
   timeout: 5000
-  database: bjc_teachers_prod
\ No newline at end of file
+  database: bjc_teachers_prod
+
+production:
```
Comment (cycomachead) [inline config/database.yml:28]: Shouldn’t this key be staging?

**Comment by Reimirno** [comment]
Will resolve by today.

File: config/database.yml (hunk 2 above)
Comment (Reimirno) [reply config/database.yml:28]: Oh I think I got prod and stage database name mixed up.

File: config/cable.yml (hunk 1 above)
Comment (Reimirno) [reply config/cable.yml:15]: Yea probably.

File: README.md [hunk 3]
```diff
@@ -75,14 +75,22 @@ We have worked on the adding following core features and functionality:
...
-  - First run `heroku pg:psql` or `psql bjc_teachers_dev`
-  - Then
-    ```
-    UPDATE teachers
-    SET admin = true
-    WHERE Email LIKE '%@berkeley.edu%'
-    ;
-    ```
+  - Using psql
+    - First run `heroku pg:psql` or `psql bjc_teachers_dev`
```
Comment (cycomachead) [inline README.md:79]: You can also just do `rails db`, locally or on heroku.

File: config/secrets.yml [hunk 4]
```diff
@@ -27,3 +27,8 @@ production:
   secret_key_base: <%= ENV["SECRET_KEY_BASE"] %>
   bjc_password: <%= ENV["BJC_PASSWORD"] %>
   piazza_password: <%= ENV["PIAZZA_PASSWORD"] %>
+
+staging:
+  secret_key_base: <%= ENV["SECRET_KEY_BASE"] %>
+  bjc_password: <%= ENV["BJC_PASSWORD"] %>
+  piazza_password: <%= ENV["PIAZZA_PASSWORD"] %>
```
Comment (cycomachead) [inline config/secrets.yml:34]:
```suggestion
  <<: *production
```

This would reduce some duplication

**Review by cycomachead — CHANGES_REQUESTED** [review]
A couple minor notes too

File: config/secrets.yml [hunk 5]
```diff
@@ -23,7 +23,10 @@ test:
 
 # Do not keep production secrets in the repository,
 # instead read values from the environment.
-production:
+production: &production
```
Comment (Reimirno) [inline config/secrets.yml:26]: `<<: *production` below will work only if we add this `&production` alias.

**Comment by Reimirno** [comment]
Everything mentioned above fixed.

I will make sure to make fewer such careless mistakes in the future.

File: config/database.yml [hunk 6]
```diff
@@ -19,6 +19,12 @@ test:
   <<: *default
   database: bjc_teachers_test
 
+production:
```
Comment (cycomachead) [inline config/database.yml:22]:
```suggestion
staging:
```

File: README.md [hunk 7]
```diff
@@ -97,25 +106,25 @@ We have worked on the adding following core features and functionality:
 ## Steps to Deploying on Heroku
 
 - ... create a heroku app
+- `heroku stack:set heroku-20` (Ruby 2.7.7 is not supported on latest stack heroku-22. Double check your Ruby version though)
```
Comment (cycomachead) [inline README.md:109]:
```suggestion
```

This should no longer be necessary with Ruby 3 as the default. :)

**Review by cycomachead — CHANGES_REQUESTED** [review]
Two super minor things.

File: README.md (hunk 7 above)
Comment (Reimirno) [reply README.md:109]: But we are using 2.7.7 though?

**Comment by cycomachead** [comment]
If you’ve merged in the upstream changes, and are on Rails 7, then it should be ruby 3 now. (Which also does work on Heroku 20 there’s no need for it explicitly)-- Michael BallFrom my iPhonemichaelball.coOn Mar 13, 2023, at 3:01 AM, Yu Long ***@***.***> wrote:﻿
@Reimirno commented on this pull request.



In README.md:
> @@ -97,25 +106,25 @@ We have worked on the adding following core features and functionality:
 ## Steps to Deploying on Heroku
 
 - ... create a heroku app
+- `heroku stack:set heroku-20` (Ruby 2.7.7 is not supported on latest stack heroku-22. Double check your Ruby version though)

But we are using 2.7.7 though?

File: README.md (hunk 7 above)
Comment (cycomachead) [reply README.md:109]: oh, ignore me! I'm getting confused. (only snap con was recently updated...the ruby 3 stuff for teacher tracker isn't done...)

**Comment by Reimirno** [comment]
All typo fixed. Merging.

**Pull request merged**